core/controllers/base_test.py

```diff
@@ -274,6 +288,132 @@ def test_renders_error_page_with_iframed(self):
             'Uh-oh! The Oppia exploration you requested may have been removed '
             'or deleted.', response.body)
 
+    def test_dev_mode_cannot_be_true_on_production(self):
+        del sys.modules['feconf']
+        server_software_swap = self.swap(
+            os, 'environ', {'SERVER_SOFTWARE': 'Production'})
+        assert_raises_regexp_context_manager = self.assertRaisesRegexp(
+            Exception, 'DEV_MODE can\'t be true on production.')
+        with assert_raises_regexp_context_manager, server_software_swap:
+            import feconf  # pylint: disable-all
+
+    def test_valid_pillow_path(self):
+        import appengine_config
+        assert_raises_regexp_context_manager = self.assertRaisesRegexp(
+            Exception, 'Invalid path for oppia_tools library: invalid_path')
+
+        def mock_os_path_join_for_pillow(*args):
```

Comment:
Add a comment/docstring explaining conceptually what this function does.

Reply:
Done